libsolidity/analysis/FunctionCallGraph.cpp

```diff
@@ -0,0 +1,149 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libsolidity/analysis/FunctionCallGraph.h>
+
+using namespace std;
+
+namespace solidity::frontend
+{
+
+
+shared_ptr<FunctionCallGraphBuilder::ContractCallGraph> const FunctionCallGraphBuilder::create(ContractDefinition const& _contract)
+{
+	m_contract = &_contract;
+
+	std::shared_ptr<ContractCallGraph> graph= make_shared<ContractCallGraph>(_contract);
+
+	// Create graph for constructor, state vars, etc
+	m_nodes = &graph->creationCalls;
+	visitConstructor(_contract);
+
+
+	// Create graph for all publicly reachable functions
+	m_nodes = &graph->runtimeCalls;
+	for (auto& [hash, functionType]: _contract.interfaceFunctionList())
+		if (auto const* funcDef = dynamic_cast<FunctionDefinition const*>(&functionType->declaration()))
+			if (!m_nodes->count(funcDef))
+				visitCallable(funcDef);
+
+	m_contract = nullptr;
+	m_nodes = nullptr;
+	solAssert(m_currentNode == nullptr, "Current node not properly reset.");
+
+	return graph;
+}
+
+bool FunctionCallGraphBuilder::visit(Identifier const& _identifier)
+{
+	if (auto const* callable = dynamic_cast<CallableDeclaration const*>(_identifier.annotation().referencedDeclaration))
+	{
+		if (*_identifier.annotation().requiredLookup == VirtualLookup::Virtual)
+			callable = &callable->resolveVirtual(*m_contract);
+
+		solAssert(
+			*_identifier.annotation().requiredLookup != VirtualLookup::Super,
+			"Only MemberAccess can have lookup 'super'"
+		);
+
+		// Ignore calls to other contract
+		if (callable->annotation().contract && !m_contract->derivesFrom(*callable->annotation().contract) && !callable->annotation().contract->isLibrary())
+			return true;
+
+		if (!m_nodes->count(callable))
+			visitCallable(callable);
+
+		m_currentNode->calls.emplace(*m_nodes->find(callable));
+	}
+
+	return true;
+}
+
+bool FunctionCallGraphBuilder::visit(NewExpression const& _newExpression)
+{
+	if (ContractType const* contractType = dynamic_cast<ContractType const*>(_newExpression.typeName().annotation().type))
+	{
+		CallableDeclaration const* callable = contractType->contractDefinition().constructor();
```

Comment:
A new expresison does not really call the constructor, but you can collect the set of contracts that is created as a separate member variable of the call graph.